numpy/_core/src/multiarray/nditer_pywrap.c

```diff
@@ -921,27 +934,54 @@ NpyIter_NestedIters(PyObject *NPY_UNUSED(self),
         Py_DECREF(item);
     }
 
-    /* op and op_flags */
-    if (npyiter_convert_ops(op_in, op_flags_in, op, op_flags, &nop)
-                                                        != 1) {
+    /* Need nop to set up workspaces */
+    PyObject **op_objs = NULL;
+    PyObject *op_in_owned;  /* Sequence/object owning op_objs. */
+    int nop = npyiter_prepare_ops(op_in, &op_in_owned, &op_objs);
+    if (nop < 0) {
         return NULL;
     }
 
-    /* Set the dtypes to all NULL to start as well */
-    memset(op_request_dtypes, 0, sizeof(op_request_dtypes[0])*nop);
-    memset(op_request_dtypes_inner, 0,
-                        sizeof(op_request_dtypes_inner[0])*nop);
+    /* allocate workspace for Python objects (operands and dtypes) */
+    NPY_ALLOC_WORKSPACE(op, PyArrayObject *, 3 * 8, 3 * nop);
+    if (op == NULL) {
+        Py_DECREF(op_in_owned);
+        return NULL;
+    }
+    memset(op, 0, sizeof(PyObject *) * 3 * nop);
+    PyArray_Descr **op_request_dtypes = (PyArray_Descr **)(op + nop);
+    PyArray_Descr **op_request_dtypes_inner = op_request_dtypes + nop;
+
+    /* And other workspaces (that do not need to clean up their content) */
```

Comment:
Same comment as above - might as well move these above `op` and write `finish` so it can handle `op == NULL`